src/lib/badge/badge.ts

```diff
@@ -0,0 +1,74 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+ import {
+   Component, Input, ViewEncapsulation, Renderer2, ElementRef, ChangeDetectionStrategy
+} from '@angular/core';
+ import {coerceBooleanProperty} from '@angular/cdk/coercion';
+ import {CanColor, mixinColor} from '@angular/material/core';
+
+ /** @docs-private */
+export class MatBadgeBase {
+  constructor(public _renderer: Renderer2, public _elementRef: ElementRef) {}
+}
+export const _MatBadgeMixinBase = mixinColor(MatBadgeBase);
+
+@Component({
+  selector: 'mat-badge',
+  templateUrl: './badge.html',
+  styleUrls: ['./badge.css'],
+  inputs: ['color'],
+  host: {
+    'class': 'mat-badge',
+    '[class.mat-badge-overlap]': 'overlap',
+    '[class.mat-badge-above]': '_isAbove',
+    '[class.mat-badge-below]': '!_isAbove',
+    '[class.mat-badge-before]': '!_isAfter',
+    '[class.mat-badge-after]': '_isAfter',
```

Comment:
I think that you can cut down on the amount of classes by making it appear `above before` by default. The other classes can override it afterwards.

Reply:
I started with that but it got odd setting everything backwards. For instance `before` would have to set `right: auto` and `left: 22px`. It felt odd to add that. Happy to change if you guys prefer.